src/functions.c

```diff
@@ -0,0 +1,445 @@
+        addReplyBulkCString(c, "engine");
+        addReplyBulkCBuffer(c, fi->ei->name, sdslen(fi->ei->name));
+        addReplyBulkCString(c, "desc");
+        if (fi->desc) {
+            addReplyBulkCBuffer(c, fi->desc, sdslen(fi->desc));
+        } else {
+            addReplyNull(c);
+        }
+    }
+    dictReleaseIterator(iter);
+}
+
+/*
+ * FUNCTION DELETE <FUNCTION NAME>
+ */
+void functionsDeleteCommand(client *c) {
+    if (server.masterhost && server.repl_slave_ro && !(c->flags & CLIENT_MASTER)) {
+        addReplyError(c, "Can not delete a function on a read only replica");
+        return;
+    }
+
+    robj *function_name = c->argv[2];
+    functionInfo *fi = dictFetchValue(functions_ctx->functions, function_name->ptr);
+    if (!fi) {
+        addReplyError(c, "Function not found");
+        return;
+    }
+
+    engineFunctionFree(fi, functions_ctx);
+    forceCommandPropagation(c, PROPAGATE_REPL | PROPAGATE_AOF);
+    addReply(c, shared.ok);
+}
+
+void functionsKillCommand(client *c) {
+    scriptKill(c, 0);
+}
+
+static void fcallCommandGeneric(client *c, int ro) {
+    robj *function_name = c->argv[1];
+    functionInfo *fi = dictFetchValue(functions_ctx->functions, function_name->ptr);
+    if (!fi) {
+        addReplyError(c, "Function not found");
+        return;
+    }
+    engine *engine = fi->ei->engine;
+
+    long long numkeys;
+    /* Get the number of arguments that are keys */
+    if (getLongLongFromObject(c->argv[2], &numkeys) != C_OK) {
+        addReplyError(c, "Bad number of keys provided");
+        return;
+    }
+    if (numkeys > (c->argc - 3)) {
+        addReplyError(c, "Number of keys can't be greater than number of args");
+        return;
+    } else if (numkeys < 0) {
+        addReplyError(c, "Number of keys can't be negative");
+        return;
+    }
+
+    scriptRunCtx r_ctx;
+
+    scriptPrepareForRun(&r_ctx, fi->ei->c, c, fi->name);
+    if (ro) {
+        r_ctx.flags |= SCRIPT_READ_ONLY;
+    }
+    engine->call(&r_ctx, engine->engine_ctx, fi->function, c->argv + 3, numkeys,
+                 c->argv + 3 + numkeys, c->argc - 3 - numkeys);
+    scriptResetRun(&r_ctx);
+}
+
+/*
+ * FCALL <FUNCTION NAME> nkeys <key1 .. keyn> <arg1 .. argn>
+ */
+void fcallCommand(client *c) {
+    fcallCommandGeneric(c, 0);
+}
+
+/*
+ * FCALL_RO <FUNCTION NAME> nkeys <key1 .. keyn> <arg1 .. argn>
+ */
+void fcallCommandReadOnly(client *c) {
+    fcallCommandGeneric(c, 1);
+}
+
+void functionsHelpCommand(client *c) {
+    const char *help[] = {
+"CREATE <ENGINE NAME> <FUNCTION NAME> [REPLACE] [DESC <FUNCTION DESCRIPTION>] <FUNCTION CODE>",
+"    Create a new function with the given function name and code.",
+"DELETE <FUNCTION NAME>",
+"    Delete the given function.",
+"INFO",
+"    For each function, print the following information about the function:",
+"    * Function name",
+"    * The engine used to run the function",
+"    * Function description",
```

Comment:
Why aren't we also printing back the raw function code? Do we think it's too much memory to store?

Reply:
We actually do store the raw function code to be able to save it to RDB. I thought that returning it on info might be to much and also not sure we want to commit to always be able to return the raw function code (it also might be meaningless on other engines in the future, like web-assembly for example). WDYT?

Reply:
@oranagra I think it will be irrelevant with WASM because the "code" we will get is a binary blob, nothing readable or meaningful.

Reply:
first, everything can be desassembled 8-)
but also, people can get it in order to load it into another redis (or another system).
i think it is ok if we spit out what we got as input.

Reply:
Its actually on the following PR list:
> Get the source code of a function
So I will come with a proposal and implementation.

Reply:
I think having a separate subcommand with more info sounds like the right approach.

Reply:
Should we do it on this PR or open an issue with API definition and then handle it on following PR?

Reply:
IMHO if you have time to kill, and it doesn't add another major approval barrier to this PR, you can go ahead and add it.

Reply:
Added the option to specify function name on FUNCTION INFO command that also return the raw code. Updated the top comment accordingly.

Reply:
does that mean that `FUNCTION INFO`, and `FUNCTION INFO <name>` return completely different result?
if it does, i don't like that.

Reply:
i think we should have
* `FUNCTION STATS` - general info on the engine and some counters (no per function details)
* `FUNCTION LIST` - listing all function and some metadata on each (excluding the function code)
* `FUNCTION INFO <name> [WITHCODE]` - detailed info on the function, optionally including it's code.

Reply:
Updated the code accordingly, also updated top comment.